### DISCUSSION TITLE
Compiling Lua code without local / global declarations?

### DISCUSSION THREAD

**Discussion by MikeDX** [edubart/nelua-lang#12] [Q&A] [answered]
Is there a command line option, or a workaround to use nelua on simple code like:

```lua
a = 1
b = 2
print(a + b)
```

rather than

```lua
local a = 1
local b = 1
print(a + b)
```

I know issues arent a wishing well, and this really isn't an issue per se, but at this point I don't know a better place to ask :)

**Comment by edubart** [comment, marked as answer]
In short no, you should explicitly declare your variable as `local` or `global`. This was supported before but I concluded that it did more harm than good for the users, because it hides away many possible errors that could be caught at compile time and this leads to bad experiences while coding. I know this is possible in Lua, but not in Nelua, furthermore many people consider bad practice and dangerous doing that even in Lua. In the future there is the possibility to revisit this feature if I decide to implement the global `_G` table like in Lua, but would be optional.

So the equivalent Nelua code for your Lua example is:
```lua
global a = 1
global b = 2
print(a + b)
```

But in general you should use local, like in Lua people usually do, unless you really want the variable accessible across modules.

> I know issues arent a wishing well, and this really isn't an issue per se, but at this point I don't know a better place to ask :)

You can ask more questions in the discord chat, the link is on the chat badge in the github home page.

**Reply by stefanos82** [reply]
> In short no, you should explicitly declare your variable as `local` or `global`. This was supported before but I concluded that it did more harm than good for the users, because it hides away many possible errors that could be caught at compile time and this leads to bad experiences while coding. I know this is possible in Lua, but not in Nelua, furthermore many people consider bad practice and dangerous doing that even in Lua. In the future there is the possibility to revisit this feature if I decide to implement the global `_G` table like in Lua, but would be optional.

That's interesting, I honestly didn't know!

Let's assume I am in a file named `stefanos.nelua` and I have declared `name: string = "stefanos"`; since I haven't explicitly declared it as `global`, why should not it be interpreted as `local` by the parser behind the scenes?

Can you help me understand the logic behind this decision?

**Reply by edubart** [reply]
There is another reason for `local` and `global` to be mandatory in the syntax, for example let's say we drop the global keyword in
```lua
global a: vector(integer)
```
the above is this a valid syntax in Nelua for declaring a dynamic array of integers, by dropping the `global` keyword we would have:
```lua
a: vector(integer) -- declaring a global
```
but spaces are really optional in Nelua (and Lua), so we could trim down to:
```lua
a:vector(integer)
```
now is that a variable declaration or a method call? the parser, or the reader wouldn't know, so a fix to the issue is to always use a keyword in variable declarations, thus let's add back the `global` keyword for variable declaration.

Other fixes would be to use another token for variable declarations, say `::`, thus maybe:
```lua
a :: vector(integer)
```
but then `::` is reserved for goto and if you declare two variables sequentially you will have syntax issues with gotos in the parser.
Ok, maybe then `:=`
```lua
a := vector(integer)
```
Then you see the syntax starts to become something else and unfamiliar, `:` is used for type notations being inspired by many languages, like Terra, Teal, Typed Lua, Titan, Nim.

**Reply by stefanos82** [reply]
Oooh...the colon! :expressionless:  I haven't thought of that scenario before!

All of a sudden, `local` looks so appealing to my eyes :rofl:

**Comment by MikeDX** [comment]
Thanks @edubart that makes sense. I'll join the discord since I'm very keen on the project! Feel free to close the issue.

**Comment by MetlHedd** [comment]
The invite link for discord server is invalid @edubart

**Comment by edubart** [comment]
@MetlHedd Thanks for notifying, I've fixed.
